mapviz_plugins/src/pose_array_plugin.cpp

```diff
@@ -0,0 +1,299 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+ * POSSIBILITY OF SUCH DAMAGE.
+ **/ 
+
+#include <mapviz_plugins/pose_array_plugin.h>
+
+// C++ standard libraries
+#include <cstdio>
+#include <vector>
+
+// QT libraries
+#include <QDialog>
+#include <QGLWidget>
+#include <QPalette>
+
+#include <opencv2/core/core.hpp>
+
+// ROS libraries
+#include <ros/master.h>
+
+#include <swri_image_util/geometry_util.h>
+#include <swri_transform_util/transform_util.h>
+#include <mapviz/select_topic_dialog.h>
+
+// Declare plugin
+#include <pluginlib/class_list_macros.h>
+PLUGINLIB_EXPORT_CLASS(mapviz_plugins::PoseArrayPlugin, mapviz::MapvizPlugin)
+
+namespace mapviz_plugins
+{
+  PoseArrayPlugin::PoseArrayPlugin() : config_widget_(new QWidget())
+  {
+    ui_.setupUi(config_widget_);
+
+    ui_.color->setColor(Qt::green);
+
+    // Set background white
+    QPalette p(config_widget_->palette());
+    p.setColor(QPalette::Background, Qt::white);
+    config_widget_->setPalette(p);
+
+    // Set status text red
+    QPalette p3(ui_.status->palette());
+    p3.setColor(QPalette::Text, Qt::red);
+    ui_.status->setPalette(p3);
+
+    QObject::connect(ui_.selecttopic, SIGNAL(clicked()), this,
+                     SLOT(SelectTopic()));
+    QObject::connect(ui_.topic, SIGNAL(editingFinished()), this,
+                     SLOT(TopicEdited()));
+    QObject::connect(ui_.positiontolerance, SIGNAL(valueChanged(double)), this,
+                     SLOT(PositionToleranceChanged(double)));
+    QObject::connect(ui_.buffersize, SIGNAL(valueChanged(int)), this,
+                     SLOT(BufferSizeChanged(int)));
+    QObject::connect(ui_.drawstyle, SIGNAL(activated(QString)), this,
+                     SLOT(SetDrawStyle(QString)));
+    QObject::connect(ui_.static_arrow_sizes, SIGNAL(clicked(bool)),
+                     this, SLOT(SetStaticArrowSizes(bool)));
+    QObject::connect(ui_.arrow_size, SIGNAL(valueChanged(int)),
+                     this, SLOT(SetArrowSize(int)));
+    QObject::connect(ui_.color, SIGNAL(colorEdited(const QColor&)), this,
+            SLOT(SetColor(const QColor&)));
+    QObject::connect(ui_.show_laps, SIGNAL(toggled(bool)), this,
+            SLOT(LapToggled(bool)));
+    QObject::connect(ui_.buttonResetBuffer, SIGNAL(pressed()), this,
+                     SLOT(ClearPoints()));
+  }
+
+  PoseArrayPlugin::~PoseArrayPlugin()
+  {
+  }
+
+  void PoseArrayPlugin::SelectTopic()
+  {
+    ros::master::TopicInfo topic =
+        mapviz::SelectTopicDialog::selectTopic("geometry_msgs/PoseArray");
+
+    if (!topic.name.empty())
+    {
+      ui_.topic->setText(QString::fromStdString(topic.name));
+      TopicEdited();
+    }
+  }
+
+  void PoseArrayPlugin::TopicEdited()
+  {
+    std::string topic = ui_.topic->text().trimmed().toStdString();
+    if (topic != topic_)
+    {
+      initialized_ = false;
```

Comment:
What is initialized_ used for? I see it assigned to, but we don't seem to check it anywhere.

Reply:
I don't see initialized_ or has_message_ being used for anything, but I'm suspicious since they are in pretty much all plugins?  Are they vestigal?

Reply:
`initialized_` is used in the base MapvizPlugin class; Mapviz will not draw a plugin until that is true, so plugins can use that to disable rendering until they're ready.

`has_message_` isn't part of any base class, but many plugins use something like that to track whether they've received any messages or not, so it may just be a copy/paste issue here.

Reply:
Turns out that this one uses it just like that (that part was just copied from pose_plugin, but not by error.)